python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -108,46 +92,82 @@ def _convert_arrow_table_to_examples(
 
 def _get_feature_value(
     feature: "tf.train.Feature",
-) -> Union[List[bytes], List[float], List[int]]:
+) -> "pyarrow.Array":
+    import pyarrow as pa
+
     values = (
-        feature.bytes_list.value,
-        feature.float_list.value,
-        feature.int64_list.value,
+        feature.HasField("int64_list"),
+        feature.HasField("float_list"),
+        feature.HasField("bytes_list"),
     )
-    # Exactly one of `bytes_list`, `float_list`, and `int64_list` should contain data.
-    assert sum(bool(value) for value in values) == 1
-
-    if feature.bytes_list.value:
-        return list(feature.bytes_list.value)
-    if feature.float_list.value:
-        return list(feature.float_list.value)
-    if feature.int64_list.value:
-        return list(feature.int64_list.value)
-
-
-def _value_to_feature(value: Union[bytes, float, int, List]) -> "tf.train.Feature":
+    # At most one of `bytes_list`, `float_list`, and `int64_list` contains data.
+    # If none contain data, this indicates an empty feature value.
+    assert sum(bool(value) for value in values) <= 1
+
+    if feature.HasField("bytes_list"):
+        value = feature.bytes_list.value
+        type_ = pa.binary()
+    elif feature.HasField("float_list"):
+        value = feature.float_list.value
+        type_ = pa.float32()
+    elif feature.HasField("int64_list"):
+        value = feature.int64_list.value
+        type_ = pa.int64()
+    else:
+        value = []
+        type_ = pa.null()
+    value = list(value)
+    if len(value) == 1:
+        # Use the value itself if the features contains a single value.
+        # This is to give better user experience when writing preprocessing UDF on
+        # these single-value lists.
+        value = value[0]
+    else:
+        # If the feature value is empty, set the type to null for now
+        # to allow pyarrow to construct a valid Array; later, infer the
+        # type from other records which have non-empty values for the feature.
+        if len(value) == 0:
+            type_ = pa.null()
```

Comment:
Hmm for this, is there a reason that the existing value type (e.g. `pa.int64()`) would work when constructing the empty `ListArray` below? It seems like `pa.list_(pa.int64())` should work instead of `pa.list_(pa.null())`, but I might be missing something.

Reply:
Oh, or do we want this to be the opaque list in case it's actually an unwrapped scalar in other blocks?

Reply:
Just saw your comment above, sounds good!

Reply:
Yeah I believe we want the type to be opaque here and infer from other non-empty records, let me know if you think it's a concern otherwise

Reply:
Yeah I believe we want the type to be opaque here and infer from other non-empty records, let me know if you think it's a concern otherwise